[WIP] Support for the NX file format

RaphaelWoude opened the pull request:
# This is a work in progress PR for adding the NX file format support.
The NX file format is a format that I would really like to support for argon as it's a very flexible and fast file format.
Note that this PR will not change the way the data is stored, it will just give you an option to load from NX instead.


## Common
- [x] NXStringDataLoader (String.nx)
- Code has been tested and works great.
- [x] NXItemDataLoader (Item.nx) 
- Code has been tested and works great.
- EvolvePet is the only node that we are missing.

## Game
- [ ] NXQuestDataLoader (Quest.nx)
- [x] NXBeautyDataLoader (Character.nx)
- Code has been tested and works great.
- [ ] NXMapDataLoader (Map.nx)
- [ ] NXMobDataLoader (Mob.nx)
- [x] NXNpcDataLoader (Npc.nx)
- Needs testing
- [ ] NXReactorDataLoader (Reactor.nx)
- [ ] NXSkillDataLoader (Skill.nx)

## Shop
- [ ] NXCashShopDataLoader (Etc.nx)
- We probably won't use this a lot but would still be great to have the option.

RaphaelWoude committed:
Added pkgnx to pom.xml

RaphaelWoude committed:
Added NX as a type to DataFileType

RaphaelWoude committed:
Added NXStringDataLoader

RaphaelWoude committed:
Finished NXStringDataLoader

RaphaelWoude committed:
Finished NXBeautyDataLoader

RaphaelWoude committed:
Added NXItemDataLoader

RaphaelWoude committed:
Added correct bytes for reqStat to StatEffect

geospiza-fortis commented on src/argonms/common/loading/item/NXItemDataLoader.java:
```diff
@@ -0,0 +1,413 @@
+                if (baseNode.getName().equals("Pet")) {
+                    for (NXNode imgNode : baseNode) {
+                        int itemId = Integer.parseInt(imgNode.getName().replace(".img", ""));
+                        doWork(itemId, imgNode);
```
`doWork` should probably be split up into several smaller private functions that are scoped to the nx file it's working on. I'd go with names like like `loadInfo`, `loadMob`, etc.

RaphaelWoude replied:
Agreed. I'll start splitting up the methods when I get home tonight.

RaphaelWoude replied:
Done
https://github.com/geospiza-fortis/argonms-server/pull/15/commits/1a875f8e2a6114e90168b507938a653ac61df363

geospiza-fortis commented on src/argonms/common/loading/item/NXItemDataLoader.java:
```diff
@@ -0,0 +1,413 @@
+        NXNode infoNode = node.getChild("info");
+
+        if (infoNode != null) {
+            short[] incStats = new short[16];
```
Where did 16 come from?

RaphaelWoude replied:
There are 16 different stats that are available for an equip. 13 of them are used and the other 3 are unused.

geospiza-fortis commented on src/argonms/common/loading/item/NXItemDataLoader.java:
```diff
@@ -0,0 +1,413 @@
+                    }
+                }
+             }
+        } catch (IOException e) {
```
Does the IOException happen outside of loading the nx file?

RaphaelWoude replied:
No. The only other exception that can be thrown after loading is while parsing with integer, short, byte etc. This will throw a NumberFormatException if it goes wrong. I'll add another catch block for this when I get home tonight. Good catch!

RaphaelWoude replied:
Done
https://github.com/geospiza-fortis/argonms-server/pull/15/commits/c4dcf7f214d2713266567d5c0cea44722fcd8c7b

geospiza-fortis commented on src/argonms/common/loading/item/NXItemDataLoader.java:
```diff
@@ -0,0 +1,413 @@
+    @Override
+    public boolean loadAll() {
+        try {
+            NXFile file = new EagerNXFile(dataPath + "Item.nx" + File.separatorChar);
```
What's with the separator char at the end?

RaphaelWoude replied:
Consistency. If you look at the KVJ loaders you can see that it always has a seperatorChar at the end of the dataPath.
Don't think it's really needed but I just wanna be consistent.

geospiza-fortis commented on src/argonms/common/loading/item/KvjItemDataLoader.java:
```diff
@@ -270,13 +267,6 @@ private void processReqStat(LittleEndianReader reader, Integer oId) {
 		reqStats.get(oId)[stat] = value;
 	}
 
-	private List<Integer> processScrollReqs(LittleEndianReader reader) {
```
It'd be nice to stage this in a separate commit instead of bundling it together with the rest of the changes.

RaphaelWoude replied:
This will be re-implemented later on in the PR because I actually found the NXNode for it.
Still will be unused and might be implemented later as a functionality when I know more about it.

RaphaelWoude replied:
Re-implemented https://github.com/geospiza-fortis/argonms-server/pull/15/commits/41c0f1e78d94ab357fc6f79ce1517a23a3337654
Done https://github.com/geospiza-fortis/argonms-server/pull/15/commits/eda49f8ca820bf9bbfadac21c75cdcc82ee73eeb

geospiza-fortis reviewed (commented):
Looks like a fairly sizable project in and of itself, I might be able to contribute a small piece sometime this week if you're going do this in chunks (maybe the map module, if any).

RaphaelWoude committed:
Split up doWork into multiple loading functions

RaphaelWoude committed:
Added catch for NumberFormatException

RaphaelWoude committed:
Added logging to NXItemDataLoader

RaphaelWoude committed:
Added logging to NXStringDataLoader

RaphaelWoude committed:
Reverted scrollRequirements

Node for this was found. Will probably still be unused.

RaphaelWoude committed:
Added loading for the requirement node

RaphaelWoude committed:
Fixed a out of bounce exception.

reqJob is not a requirement in the 16 bytes. There is no handling for this yet.

RaphaelWoude committed:
Added loading for petConsumableBy

RaphaelWoude committed:
Removed endEffect as NX does not support it.

End effect is a property created by the KVJ format to say that there are no more effects to be added to an item.

RaphaelWoude committed:
FIxed logger name for NXItemDataLoader

RaphaelWoude committed:
Added NXNpcDataLoader

RaphaelWoude committed:
Moved directory into it's own variable

geospiza-fortis commented on src/argonms/game/loading/npc/NXNpcDataLoader.java:
```diff
@@ -0,0 +1,91 @@
+        }
+    }
+
+    private NXNode getNode(int npcId) {
```
I'd annotate this as the source of the NumberFormatException.

geospiza-fortis reviewed (commented):
The new changes look reasonable 👍
